### PR TITLE
Add MetricsQueryResult to MonitorQueryModelFactory in Azure.Monitor.Query

### PR DESCRIPTION
**Pull request by nisha-bhatia**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by nisha-bhatia** [comment]
Addresses: https://github.com/Azure/azure-sdk-for-net/issues/25299

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs [hunk 1]
```diff
@@ -60,5 +63,45 @@ public void ExposesPublicEndpoint()
             var client = new MetricsQueryClient(new Uri("https://management.azure.gov"), new MockCredential(), new MetricsQueryClientOptions());
             Assert.AreEqual(new Uri("https://management.azure.gov"), client.Endpoint);
         }
+
+        [Test]
+        public void CanGetMetricQueryResult()
+        {
+            MetricTimeSeriesElement metricTimeSeriesElement = new Models.MetricTimeSeriesElement();
+            Assert.IsNotNull(metricTimeSeriesElement);
+            Assert.AreEqual(0, metricTimeSeriesElement.Metadata.Count);
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:72]: can you add some Metadata and Values to this example?

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs [hunk 2]
```diff
@@ -60,5 +63,45 @@ public void ExposesPublicEndpoint()
...
+            Assert.AreEqual(0, metricTimeSeriesElement.Values.Count);
+            IEnumerable<MetricTimeSeriesElement> metricTimeSeriesElements = new[] { metricTimeSeriesElement };
+
+            MetricUnit metricUnit = new MetricUnit("test");
+            Assert.IsNotNull(metricUnit);
+            Assert.AreEqual("test", metricUnit.ToString());
+
+            MetricResult metricResult = new Models.MetricResult("https://management.azure.gov", "type", new LocalizableString("name"), metricUnit, metricTimeSeriesElements);
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:80]: Is this MetricResult constructor public?

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs [hunk 3]
```diff
@@ -60,5 +63,45 @@ public void ExposesPublicEndpoint()
...
+            Assert.AreEqual(null, metricResult.Error.Code);
+            Assert.AreEqual(null, metricResult.Error.Message);
+            Assert.AreEqual("https://management.azure.gov", metricResult.Id);
+            Assert.AreEqual("name", metricResult.Name);
+            Assert.AreEqual("type", metricResult.ResourceType);
+            Assert.AreEqual(1, metricResult.TimeSeries.Count);
+            Assert.AreEqual("test", metricResult.Unit.ToString());
+            IEnumerable<MetricResult> metricResults = new[] { metricResult };
+
+            MetricsQueryResult metricsQueryResult = new Models.MetricsQueryResult(null, TimeSpan.FromMinutes(3).ToString(), null, "namespace", "eastus", metricResults.ToList());
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:92]: Is this MetricsQueryResult constructor public?

Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:92]: yes

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs (hunk 2 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:80]: yes

Comment (pakrym) [reply sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:80]: Looks like it's internal: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.Query/src/Generated/Models/MetricResult.cs

**Review by scottaddie — CHANGES_REQUESTED** [review]

File: sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs [hunk 4]
```diff
@@ -22,10 +22,31 @@ public static partial class MonitorQueryModelFactory
...
-            return new MetricsQueryResult(cost, timespan, granularity, @namespace, resourceRegion, metrics);
+            return new MetricsQueryResult(cost, timespan, granularity, @namespace, resourceRegion, metrics.ToArray());
+        }
+
+        /// <summary>
+        /// Enables the user to create an instance of a <see cref="MetricResult"/>.
+        /// </summary>
+        /// <param name="id"> The metric Id. </param>
+        /// <param name="resourceType"> The resource type of the metric resource. </param>
+        /// <param name="localizedName"> The name and the display name of the metric, i.e. it is localizable string. </param>
```
[the remark this answers is not in the document]
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs]: I updated the parameter name so hopefully it's more clear now

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs [hunk 5]
```diff
@@ -60,5 +63,54 @@ public void ExposesPublicEndpoint()
             var client = new MetricsQueryClient(new Uri("https://management.azure.gov"), new MockCredential(), new MetricsQueryClientOptions());
             Assert.AreEqual(new Uri("https://management.azure.gov"), client.Endpoint);
         }
+
+        [Test]
+        public void CanGetMetricQueryResult()
+        {
+            var metadata = new Dictionary<string, string> { { "metadatatest1", "metadatatest2" } };
+            var metricValue = new MetricValue(new DateTimeOffset(new DateTime(10)));
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:71]: Is this ctor public?

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs [hunk 6]
```diff
@@ -60,5 +63,54 @@ public void ExposesPublicEndpoint()
...
+            Assert.AreEqual(1, metricTimeSeriesElement.Values.Count);
+            Assert.AreEqual("TimeStamp: 1/1/0001 12:00:00 AM -08:00 ", metricTimeSeriesElement.Values[0].ToString());
+            Assert.AreEqual(new DateTimeOffset(new DateTime(10)), metricTimeSeriesElement.Values[0].TimeStamp);
+            IEnumerable<MetricTimeSeriesElement> metricTimeSeriesElements = new[] { metricTimeSeriesElement };
+
+            MetricUnit metricUnit = new MetricUnit("test");
+            Assert.IsNotNull(metricUnit);
+            Assert.AreEqual("test", metricUnit.ToString());
+
+            LocalizableString localizableString = new LocalizableString("name", "localizedName");
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:88]: This type is internal, customers can't use it.

File: sdk/monitor/Azure.Monitor.Query/api/Azure.Monitor.Query.netstandard2.0.cs [hunk 7]
```diff
@@ -347,7 +347,9 @@ public static partial class MonitorQueryModelFactory
...
         public static Azure.Monitor.Query.Models.MetricAvailability MetricAvailability(System.TimeSpan? granularity = default(System.TimeSpan?), System.TimeSpan? retention = default(System.TimeSpan?)) { throw null; }
+        public static Azure.Monitor.Query.Models.MetricResult MetricResult(string id, string resourceType, string name, Azure.Monitor.Query.Models.MetricUnit unit, System.Collections.Generic.IEnumerable<Azure.Monitor.Query.Models.MetricTimeSeriesElement> timeSeries) { throw null; }
         public static Azure.Monitor.Query.Models.MetricsQueryResult MetricsQueryResult(int? cost, string timespan, System.TimeSpan? granularity, string @namespace, string resourceRegion, System.Collections.Generic.IReadOnlyList<Azure.Monitor.Query.Models.MetricResult> metrics) { throw null; }
+        public static Azure.Monitor.Query.Models.MetricTimeSeriesElement MetricTimeSeriesElement(System.Collections.Generic.IReadOnlyDictionary<string, string> metadatavalues, System.Collections.Generic.IEnumerable<Azure.Monitor.Query.Models.MetricValue> values) { throw null; }
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/api/Azure.Monitor.Query.netstandard2.0.cs:352]: metadataValues

File: sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs (hunk 5 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Query/tests/MetricsQueryClientTests.cs:71]: sorry, yes I kept updating the Generated file by mistake. I believe I modified the correct one now.

File: sdk/monitor/Azure.Monitor.Query/src/Models/MetricValue.cs [hunk 8]
```diff
@@ -9,6 +10,23 @@ public partial class MetricValue
     {
         private string _toString;
 
+        /// <summary> Initializes a new instance of MetricValue. </summary>
+        /// <param name="timeStamp"> the timestamp for the metric value in ISO 8601 format. </param>
+        /// <param name="average"> the average value in the time range. </param>
+        /// <param name="minimum"> the least value in the time range. </param>
+        /// <param name="maximum"> the greatest value in the time range. </param>
+        /// <param name="total"> the sum of all of the values in the time range. </param>
+        /// <param name="count"> the number of samples in the time range. Can be used to determine the number of values that contributed to the average value. </param>
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Query/src/Models/MetricValue.cs:19]:
```suggestion
        /// <summary> Initializes a new instance of <see cref="MetricValue"/>. </summary>
        /// <param name="timeStamp"> The timestamp for the metric value in ISO 8601 format. </param>
        /// <param name="average"> The average value in the time range. </param>
        /// <param name="minimum"> The lowest value in the time range. </param>
        /// <param name="maximum"> The highest value in the time range. </param>
        /// <param name="total"> The sum of all values in the time range. </param>
        /// <param name="count"> The number of samples in the time range. Can be used to determine the number of values that contributed to the average value. </param>
```

File: sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs [hunk 9]
```diff
@@ -22,10 +22,39 @@ public static partial class MonitorQueryModelFactory
...
+        /// <param name="timeSeries"> The time series returned when a data query is performed. </param>
+        public static MetricResult MetricResult(string id, string resourceType, string name, MetricUnit unit, IEnumerable<MetricTimeSeriesElement> timeSeries)
+        {
+            return new MetricResult(id, resourceType, new LocalizableString(name), unit, timeSeries);
+        }
+
+        /// <summary>
+        /// Enables the user to create an instance of a <see cref="MetricTimeSeriesElement"/>.
+        /// <param name="metadataValues"> A dictionary comprised of metric metadata values. </param>
+        /// <param name="values"> A list made of MetricValue elements. </param>
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs:47]:
```suggestion
        /// <param name="values"> A list of <see cref="MetricValue"/> elements. </param>
```

File: sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs [hunk 10]
```diff
@@ -22,10 +22,39 @@ public static partial class MonitorQueryModelFactory
...
         public static MetricsQueryResult MetricsQueryResult(int? cost, string timespan, TimeSpan? granularity, string @namespace, string resourceRegion, IReadOnlyList<MetricResult> metrics)
         {
-            return new MetricsQueryResult(cost, timespan, granularity, @namespace, resourceRegion, metrics);
+            return new MetricsQueryResult(cost, timespan, granularity, @namespace, resourceRegion, metrics.ToArray());
+        }
+
+        /// <summary>
+        /// Enables the user to create an instance of a <see cref="MetricResult"/>.
+        /// </summary>
+        /// <param name="id"> The metric Id. </param>
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Query/src/Models/MonitorQueryModelFactory.cs:34]:
```suggestion
        /// <param name="id"> The metric ID. </param>
```

**Review by scottaddie — APPROVED** [review]
Here are some nits to consider, and then this looks good to me!

File: sdk/monitor/Azure.Monitor.Query/src/Models/MetricValue.cs [hunk 11]
```diff
@@ -9,6 +10,23 @@ public partial class MetricValue
...
+        public MetricValue(DateTimeOffset timeStamp, double? average, double? minimum, double? maximum, double? total, double? count)
```
Comment (pakrym) [inline sdk/monitor/Azure.Monitor.Query/src/Models/MetricValue.cs:20]: Why make the ctor public VS adding a factory method like we did for everything else?

Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Query/src/Models/MetricValue.cs]: updated

**Review by pakrym — APPROVED** [review]

**Pull request merged**